### PR TITLE
.github/workflows/rebase.yml: rebase for multiple branches

### PR DESCRIPTION
**Pull request by zowoq**
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by zowoq** [comment]
Leaving this as a draft but it's ready for review. I'll rename the action when merging.

@cdepillabout You may find this useful for PRs that need rebasing onto `haskell-updates`?

@mweinelt IIRC you asked about using it for release branches?

File: .github/workflows/rebase-staging.yml [hunk 1]
```diff
@@ -3,13 +3,13 @@ on:
...
-# This action allows people with write access to the repo to rebase a PRs base branch from
-# master to staging by commenting `/rebase-staging` on the PR while avoiding CODEOWNER notifications.
+# This action allows people with write access to the repo to rebase a PRs base branch
+# by commenting `/rebase ${branch}` on the PR while avoiding CODEOWNER notifications.
 
 jobs:
   rebase:
     runs-on: ubuntu-latest
-    if: github.repository_owner == 'NixOS' && github.event.issue.pull_request != '' && contains(github.event.comment.body, '/rebase-staging')
+    if: github.repository_owner == 'NixOS' && github.event.issue.pull_request != '' && contains(github.event.comment.body, '/rebase')
```
Comment (Mic92) [inline .github/workflows/rebase-staging.yml:12]: Should be even backwards compatible.

**Review by Mic92 — APPROVED** [review]

File: .github/workflows/rebase-staging.yml [hunk 2]
```diff
@@ -21,40 +21,103 @@ jobs:
...
+          exit 1
+        if: "! steps.check-write-access.outputs.has-permission"
+      - name: setup
+        run: |
+          curl "https://api.github.com/repos/${{ github.repository }}/pulls/${{ github.event.issue.number }}" 2>/dev/null >pr.json
+          cat <<EOF >>"$GITHUB_ENV"
+          COMMITS=$(jq -r '.commits' pr.json)
+          CURRENT_BASE=$(jq -r '.base.ref' pr.json)
+          PR_BRANCH=$(jq -r '.head.ref' pr.json)
+          COMMENT_BRANCH=$(echo ${{ github.event.comment.body }} | grep '/rebase' | awk '{print $2}')
```
Comment (mweinelt) [inline .github/workflows/rebase-staging.yml:36]: This looks a bit fragile. Maybe only accept lines beginning with `/rebase`?

```suggestion
          COMMENT_BRANCH=$(echo ${{ github.event.comment.body }} | grep -E '^/rebase' | awk '{print $2}')
```

**Review by mweinelt — COMMENTED** [review]
Loving the idea, thank you for making it happen!

**Review by mweinelt — APPROVED** [review]
Looks pretty good IMO. I don't have much experience with GHA though.

File: .github/workflows/rebase-staging.yml (hunk 1 above)
Comment (mweinelt) [reply .github/workflows/rebase-staging.yml:12]: It's not. We're expecting `/rebase <branch>` with this (see the `awk '{ print $2 }'`), while previously we accepted `/rebase-staging`.

**Comment by cdepillabout** [comment]
@zowoq Thanks for pinging me on this.  This does look like it would be helpful for rebasing stuff on `haskell-updates`.

Although I'm not very familiar with these `/` commands using GitHub actions.  

Does the `/rebase` command actually rebase a branch in a user's repo (and not just change the base branch on GitHub)?

I  was under the impression that even people with the commit bit in a repo aren't able to rebase a branch in a user's repo (even if they checked the box `allow maintainer to push to my branch` or whatever).  So I'm somewhat surprised that this functionality is available as a `/` command.

**Comment by SuperSandro2000** [comment]
> Does the /rebase command actually rebase a branch in a user's repo (and not just change the base branch on GitHub)?

Yes. Base branch can be changed on the website but the two actions need to happen at once or you ping a lot of people.

I used it multiple times already and it mostly worked.

**Comment by zowoq** [comment]
Once we've used this for a couple of weeks without any problems I'll see about getting a token added to the repo so we can drop the `close/reopen` hack.

I've added a check for `Allow edits .. by maintainers` / `maintainer_can_modify` as it seems like it might be a common error, some of the other failures look like problems with the `gh` tool itself or the runners environment.

**Comment by zowoq** [comment]
I just triggered this by accident by commenting on a PR with `Updated{slash}rebased on master`, we may need to change it to use something a bit more specific.

**Comment by mweinelt** [comment]
I tried rebasing on the same branch and for some reason that is prohibited.

> Can't rebase 20.09/microcodeIntel from release-20.09 onto release-20.09 (PR:113521 COMMITS:1)

Rebasing became necessary when an eval error hit release-20.09 and allowing rebase onto the same base would've been useful.

**Comment by zowoq** [comment]
It's disabled it to simplify the action as the merge-base and change base API commands aren't needed in that case.

The action is intended to cover the more complicated cases that also require additional steps: an API call or manually changing the base branch.

TBH I don't think it's worthwhile having support for the same branch as the action only works for committers and it's simple (and quicker) to rebase a branch on HEAD as it only needs `git` commands.

**Comment by mweinelt** [comment]
The API calls could be conditional on whether the branches aren't the same, no?

I know this is an afterthought, but It would be an amazing convenience feature, allowing the bot to take care of rebasing to HEAD to let the CI reevaluate with fixes that are on the target branch.

The benefit of the command is that it is lazy, it can be issued when you see the necessity and you don't have to open a shell, fetch and checkout the relevant branch. I believe there is little harm in allowing this.

Granted the whole action is getting quite complex, but we're talking about incremental improvements here.

**Comment by zowoq** [comment]
The confusion about how to "properly" rebase across branches without the CODEOWNER spam has generated many discussion and issues. Even the IRC bot has a programmed response. It's worthwhile having the action to simplify it.

I don't see rebasing on HEAD it as an incremental improvement, it's an increase in complexity to address a use case that is already uncomplicated.

**Comment by cdepillabout** [comment]
> I don't see rebasing on HEAD it as an incremental improvement, it's an increase in complexity to address a use case that is already uncomplicated.

While I agree that increasing complexity is definitely something to watch out for, it would be great to be able to do more directly through the github UI.

For instance, in https://github.com/NixOS/nixpkgs/pull/114418#issuecomment-786413882, I tried to use the /rebase command to do a rebase on the same branch, but it failed.  I had to do it through the CLI.

If this would work to do a rebase on the same base branch, then I could have merged in that PR without ever going to the CLI.  This would have been slightly easier for me.

**Comment by zowoq** [comment]
Okay, there's been a few of requests now for rebasing on the same branch, let's add it.

Currently it doesn't work with the close/reopen hack, we won't be able to reopen the PR because of "branch was force pushed or recreated". If we change the action to use a different token we avoid the need for close/reopen hack so that seems to be the easier option. I'll try to see about organising that in the next couple of days.

**Comment by zowoq** [comment]
> ... next couple of days.

Took a bit longer than I planned but I've opened a draft at #124560, also includes some other changes which will make the commands more reliable.

**Comment by zowoq** [comment]
PR to remove: https://github.com/NixOS/nixpkgs/pull/126192

**Pull request merged**